### PR TITLE
[FLINK-7411][network] minor performance improvements in NettyMessage

### PR DESCRIPTION
**Pull request by NicoK**
## What is the purpose of the change

This PR adds some (minor) performance improvements to `NettyMessage` which I came across: using a `switch` rather than multiple `if...ifelse...ifelse...` avoiding unneeded virtual method calls.

## Brief change log

- use a switch rather than multiple if conditions
- use static `readFrom` methods to create instances of the message sub types

## Verifying this change

This change is already covered by existing tests, such as `NettyMessageSerializationTest` as well as many other tests, especially IT cases, involving network.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes, if you consider network communication part of this)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 1]
```diff
@@ -242,21 +239,22 @@ ByteBuf write(ByteBufAllocator allocator) throws IOException {
...
-		void readFrom(ByteBuf buffer) {
-			receiverId = InputChannelID.fromByteBuf(buffer);
-			sequenceNumber = buffer.readInt();
-			isBuffer = buffer.readBoolean();
-			size = buffer.readInt();
-
-			retainedSlice = buffer.readSlice(size);
-			retainedSlice.retain();
+		static BufferResponse readFrom(ByteBuf buffer) {
+			BufferResponse result = new BufferResponse();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: this is somehow strange construct, with this default empty constructor and manually modifying fields afterwards. Why not:
```
InputChannelID receiverId = InputChannelID.fromByteBuf(buffer);
int sequenceNumber = buffer.readInt();
boolean isBuffer = buffer.readBoolean();
int size = buffer.readInt();

return new BufferResponse(
  receivedId,
  sequenceNumber,
  isBuffer,
  size,
  buffer.readSlice(size).retain());
```

would be more natural, would be easier for the finding usages of the fields and that way you could make all of the fields `final`, which is nice feature on its own.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 2]
```diff
@@ -313,20 +308,20 @@ ByteBuf write(ByteBufAllocator allocator) throws IOException {
...
 				if (!(obj instanceof Throwable)) {
 					throw new ClassCastException("Read object expected to be of type Throwable, " +
 							"actual type is " + obj.getClass() + ".");
 				} else {
-					cause = (Throwable) obj;
+					ErrorResponse result = new ErrorResponse((Throwable) obj);
 
 					if (buffer.readBoolean()) {
-						receiverId = InputChannelID.fromByteBuf(buffer);
+						result.receiverId = InputChannelID.fromByteBuf(buffer);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: ditto

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 3]
```diff
@@ -378,11 +373,14 @@ ByteBuf write(ByteBufAllocator allocator) throws IOException {
 			}
 		}
 
-		@Override
-		public void readFrom(ByteBuf buffer) {
-			partitionId = new ResultPartitionID(IntermediateResultPartitionID.fromByteBuf(buffer), ExecutionAttemptID.fromByteBuf(buffer));
-			queueIndex = buffer.readInt();
-			receiverId = InputChannelID.fromByteBuf(buffer);
+		static PartitionRequest readFrom(ByteBuf buffer) {
+			PartitionRequest result = new PartitionRequest();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: ditto

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 4]
```diff
@@ -439,20 +437,23 @@ ByteBuf write(ByteBufAllocator allocator) throws IOException {
 			}
 		}
 
-		@Override
-		public void readFrom(ByteBuf buffer) throws IOException {
+		static TaskEventRequest readFrom(ByteBuf buffer, ClassLoader classLoader) throws IOException {
+			TaskEventRequest result = new TaskEventRequest();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: ditto

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 5]
```diff
@@ -220,6 +216,7 @@ void releaseBuffer() {
 
 		@Override
 		ByteBuf write(ByteBufAllocator allocator) throws IOException {
+			assert buffer != null; // see BufferResponse()
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: `checkState`

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java [hunk 6]
```diff
@@ -130,34 +128,31 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out)
...
+					decodedMsg = ErrorResponse.readFrom(msg);
+					break;
+				case CancelPartitionRequest.ID:
+					decodedMsg = CancelPartitionRequest.readFrom(msg);
+					break;
+				case CloseRequest.ID:
+					decodedMsg = CloseRequest.readFrom(msg);
+					break;
+				default:
+					throw new IllegalStateException("Received unknown message from producer: " + msg);
```
[the remark this answers is not in the document]
Comment (NicoK) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: I agree that `IllegalStateException` is not the best fit here, but also, `UnsupportedOperationException` is not - what do you think about `ProtocolException`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java (hunk 1 above)
Comment (NicoK) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java]: I agree we could change that, too - and partly I did in another commit that didn't go into a PR yet. Let's do this for all classes then.

**Comment by NicoK** [comment]
how about now?

**Comment by tillrohrmann** [comment]
Thanks for your contribution @NicoK and the review @pnowojski. Changes look good to me. Will wait for Travis and then merge this PR.

**Pull request closed without merging**